### PR TITLE
[Reviewer: Ellie] Better management of bootstrapping cassandra

### PR DESCRIPTION
**Pull request by johadalin**
I'm hoping that the comments and code are pretty self explanatory, but i'm sure there's going to be edge cases/window conditions/weird junk i've missed.
I think the most important bit is that we don't kill cassandra when we've kicked it off, otherwise we break the bootstrapping, and are then dead in the water.

I'm also not certain on the safety of writing the yaml and topology files on the second loop into this (i.e. if the cluster manager died while waiting for cassandra). I will probably shift that so that we do nothing if the bootstrapping flag is set.

Let me know what you think.

### CODE REVIEW

File: clearwater_cassandra/cassandra_plugin.py [hunk 1]
```diff
@@ -229,6 +251,11 @@ def leave_cassandra_cluster(self):
 
         run_command("nodetool decommission", self._sig_namespace)
 
+        # Remove the bootstrapped flag so that we bootstrap correctly
+        # if rejoining the cluster again in future.
+        if os.path.exists(self.BOOTSTRAPPED_FLAG):
```
Comment (eleanor-merry) [inline clearwater_cassandra/cassandra_plugin.py]: Remove the bootstrapping in progress file too. You probably should also remove these in clearwater-cassandra's postrm.

File: clearwater_cassandra/cassandra_plugin.py [hunk 2]
```diff
@@ -163,21 +166,40 @@ def write_new_cassandra_config(self, seeds_list, destructive_restart=False):
...
+        # We do not want to kill cassandra if it is in the process of bootstrapping
+        if not os.path.exists(self.BOOTSTRAP_IN_PROGRESS_FLAG):
+            run_command("start-stop-daemon -K -p /var/run/cassandra/cassandra.pid -R TERM/30/KILL/5")
+            _log.info("Stopped Cassandra while changing config files")
+
+        # We only want to perform these steps the first time we join a cluster
+        # If we are bootstrapping, or already bootstrapped, doing this will leave
+        # us unable to rejoin the cluster properly
+        if (destructive_restart
+        and not (os.path.exists(self.BOOTSTRAPPED_FLAG)
```
Comment (eleanor-merry) [inline clearwater_cassandra/cassandra_plugin.py]: Weird indentation. Also, please bracket each part.

File: clearwater_cassandra/cassandra_plugin.py [hunk 3]
```diff
@@ -53,6 +53,9 @@ class CassandraPlugin(SynchroniserPluginBase):
     CASSANDRA_YAML_FILE = "/etc/cassandra/cassandra.yaml"
     CASSANDRA_TOPOLOGY_FILE = "/etc/cassandra/cassandra-rackdc.properties"
 
+    BOOTSTRAP_IN_PROGRESS_FLAG = "/etc/clearwater/bootstrap_in_progress"
```
Comment (eleanor-merry) [inline clearwater_cassandra/cassandra_plugin.py]: Can the file names include cassandra?

**Review by eleanor-merry — APPROVED** [review]
Code looks good, minor comments inline. Also - I assume this is going to need some UTs?

**Comment by johadalin** [comment]
Adding a postrm script here:
https://github.com/Metaswitch/clearwater-cassandra/pull/129

**Pull request merged**